signac/core/file_buffered_synced_collection.py

```diff
@@ -0,0 +1,256 @@
+        """Get the total amount of data currently stored in the buffer."""
+        return FileBufferedCollection._CURRENT_BUFFER_SIZE
+
+    @staticmethod
+    def _hash(blob):
+        """Calculate and return the md5 hash value for the file data.
+
+        Parameters
+        ----------
+        blob : bytes
+            Byte literal to be hashed.
+
+        Returns
+        -------
+        str
+            The md5 hash of the input bytes.
+        """
+        if blob is not None:
+            m = hashlib.md5()
+            m.update(blob)
+            return m.hexdigest()
+
+    def _get_file_metadata(self):
+        """Return metadata of file.
+
+        Returns
+        -------
+        Tuple[int, float]
+            The size and last modification time of the associated file.
+        """
+        try:
+            metadata = os.stat(self._filename)
+            return metadata.st_size, metadata.st_mtime
+        except OSError as error:
+            if error.errno != errno.ENOENT:
+                raise
+
+    def _flush(self, force=False):
+        """Save buffered changes to the underlying file.
+
+        Parameters
+        ----------
+        force : bool
+            If True, force a flush even in buffered mode (defaults to False).
+        """
+        if not self._is_buffered or force:
+            try:
+                cached_data = self._cache[self._filename]
+            except KeyError:
+                # There are valid reasons for nothing to be in the cache (the
+                # object was never actually accessed during global buffering,
+                # multiple collections pointing to the same file, etc).
+                pass
+            else:
+                blob = json.dumps(self.to_base()).encode()
+
+                # If the contents have not been changed since the initial read,
+                # we don't need to rewrite it.
+                if self._hash(blob) != cached_data["contents"]:
+                    # Validate that the file hasn't been changed by something
+                    # else.
+                    if cached_data["metadata"] != self._get_file_metadata():
+                        raise MetadataError(self._filename)
+                    self._data = json.loads(cached_data["contents"])
+                    self._sync()
+                del self._cache[self._filename]
+                data_size = sys.getsizeof(cached_data)
+                FileBufferedCollection._CURRENT_BUFFER_SIZE -= data_size
+
+    def _encode(self):
+        """Encode the data into a serializable form.
+
+        This method assumes JSON-serializable data, but is exposed to allow
+        changing this behavior.
+
+        Returns
+        -------
+        bytes
+            The underlying encoded data.
+        """
+        return json.dumps(self.to_base()).encode()
+
+    @staticmethod
+    def _decode(blob):
+        """Decode serialized data.
+
+        Mirroring _encode, this method assumes JSON serialization.
```

Comment:
I would copy the docstring from the previous method instead of vaguely referring to it here.

Reply:
Will do, I'm going to leave this conversation unresolved until we come to a consensus on the proper choice/documentation of the data/encoding restrictions of this class.